TNT Changes

KingCol13 opened the pull request:
Mum said it was my turn to play with the TNT code.

- Make TNT explosions always drop pickups (1.14 behaviour)
- Always drop items that wiki says should always drop
- Get rid of destination and just keep direction and intensity (seems to be wiki behaviour)
- Give each ray in explosion different random intensity

bearbin reviewed (commented):
Looks OK, but I'll let @tigerw give this a proper review.

It would be nice to have a proper way to configure the explodinator wrt blocks being thrown around, and the probabilities etc - but I'm not sure if this PR is the right place for that.

tigerw commented:
Haven't forgotten this

tigerw commented:
@KingCol13 the tracing strategy was changed from YP/YM as a pair & 4 sides as a quadruple to YP/YM, ZP/ZM, XP/XM as pairs, if I'm reading this right. Did you see benefits to using the latter?

KingCol13 commented:
@tigerw yep I changed it, I couldn't understand the YP/YM and 4 side system despite the ASCII diagrams and I didn't see any performance impact in changing. To me, each pair of sides having it's own loop seems a little easier to understand, but that's fairly subjective.

I also changed the implementation to use a 2x2 cube centred on the origin as in this diagram: https://minecraft.gamepedia.com/File:ExplosionRay.png, but again this change doesn't change the outcome.

tigerw commented on src/Physics/Explodinator.cpp:
```diff
@@ -214,59 +232,48 @@ namespace Explodinator
 		}
 	}
 
 	/** Sends out Explosion Lazors (tm) originating from the given position that destroy blocks. */
 	static void DamageBlocks(cChunk & a_Chunk, const Vector3f a_Position, const unsigned a_Power, const bool a_Fiery, const cEntity * const a_ExplodingEntity)
 	{
-		const auto Intensity = a_Power * (0.7f + GetRandomProvider().RandReal(0.6f));
-		const auto ExplosionRadius = CeilC((Intensity / StepAttenuation) * StepUnit);
 
 		// Oh boy... Better hope you have a hot cache, 'cos this little manoeuvre's gonna cost us 1352 raytraces in one tick...
-		const int HalfSide = TraceCubeSideLength / 2;
 
 		// The following loops implement the tracing algorithm described in http://minecraft.gamepedia.com/Explosion
 
 		// Trace rays from the explosion centre to all points in a square of area TraceCubeSideLength * TraceCubeSideLength
-		// in the YM and YP directions:
-		for (int OffsetX = -HalfSide; OffsetX < HalfSide; OffsetX++)
+		// Careful to avoid duplicates along edges
+		// Top and bottom sides:
+		for (float OffsetX = -1; OffsetX < 1; OffsetX+=GridPointSpacing)
```
As far as I can tell vanilla behaviour is using `int`s here so we should keep it as before

Titaniumtown commented:
About that 1.14 tnt item dropping behavior, isn't Cuberite now going to try to focus on older features first? https://github.com/cuberite/cuberite/issues/4888#issuecomment-704315535

mathiascode commented:
In my opinion, forcing (volunteer) contributors to do specific things is a good way to get rid of them. Of course you should listen to users as much as possible, but I feel most open source projects just don't have the liberty of being too picky.

tigerw commented:
I have returned! With measurements of the performances.

Off the cuff test, running DamageBlocks 10000 times with a fixed world:

1. PR verbatim: 140s
2. master: 124s
3. This PR, but with integer offsets: 119s

If I had to explain the numbers I'd say using floating points in the loop counters is slower. Separating the tracing loops vs having front/back/left/right done all at once looks like margin-of-error stuff, maybe the former helps compiler optimisation a little, I don't know.

I agree the tracing as it stands in master is unclear, even with the diagrams. @KingCol13 any objections to me changing the offset to int and merging?

tigerw commented:
About the features, as I understand there are two points of consideration; one is the version a given MC feature was introduced, another is version of the feature's behaviour we implement.

For example, explosions (super early feature), and whether explosions always drop pickups (1.14 apparently).

The linked discussion was about focusing on implementing features in ascending order of introduction to avoid a piecemeal experience of supported vs unsupported features. On the other hand, from what I can tell our policy of nevertheless implementing the latest version's behaviour, when a feature is added, is still the same.

KingCol13 commented:
@tigerw Thanks for doing the proper tests! I’d just assumed the loop increments would be negligible compared to the raytracing but the numbers don’t lie and this is definitely a sizeable speedup.

No objections at all, go ahead with the ints and merge. Cheers for the help.
